### DISCUSSION TITLE
numpy errors

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [python/typing#1206] [Q&A]
The mypy.ini file includes:

`plugins = numpy.typing.mypy_plugin`

On python+numpy code, mypy throws errors of the form:

`error: Incompatible types in assignment (expression has type "Union[Any, _SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]", variable has type "int")`

For example, the line:

`indices: np.ndarray = this_vector.nonzero()[0]`

causes the above error. If changed to a list:

`indices: list[float] = this_vector.nonzero()[0].tolist()`

it works but defeats the purpose of using numpy which performers faster and the code is cleaner.

Please explain what is going on. Thanks

**Comment by BvB93** [comment]
Can you provide an actual fully self-contained example?
The error message suggests that you're using [`npt.ArrayLike`](https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike) for an unintended purpose, but with the currently provided code snippets I can't make out anything more conclusive.

**Comment by dineshbvadhia** [comment]
Sure:
```
import numpy as np
import numpy.typing as npt

def get_indices(v: npt.ArrayLike) -> np.ndarray:
    idx: np.ndarray = v.nonzero()[0]
    return idx

def main():
    vec: np.ndarray = np.asarray([1.0, 0, 5.0, 7.0, 0], dtype=np.float32)
    print('vector:', type(vec), vec)

    indices: np.ndarray = get_indices(vec)
    print('indices:', type(indices), indices)

if __name__ == '__main__':
    main()
```
```
> python bad_np.py
vector: <class 'numpy.ndarray'> [1. 0. 5. 7. 0.]
indices: <class 'numpy.ndarray'> [0 2 3]
```
```
> mypy bad_np.py --no-incremental
bad_np.py:5: error: Item "_SupportsArray[dtype[Any]]" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "_NestedSequence[_SupportsArray[dtype[Any]]]" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "bool" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "int" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "float" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "complex" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "str" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "bytes" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
bad_np.py:5: error: Item "_NestedSequence[Union[bool, int, float, complex, str, bytes]]" of "Union[_SupportsArray[dtype[Any]], _NestedSequence[_SupportsArray[dtype[Any]]], bool, int, float, complex, str, bytes, _NestedSequence[Union[bool, int, float, complex, str, bytes]]]" has no attribute "nonzero"
Found 9 errors in 1 file (checked 1 source file)
```

**Comment by BvB93** [comment]
> ``` python
> def get_indices(v: npt.ArrayLike) -> np.ndarray:
>     idx: np.ndarray = v.nonzero()[0]
>     return idx
> ```
This explains a lot. `npt.ArrayLike` is a union of all types that can safely(-ish) be converted into an `np.ndarray`, including things like scalars and nested lists; it does _not_ represent some type of duck-array.

The error is correct in this case, what you'd want here is to either call the `np.nonzero` _function_ (not the method) or to swap the `npt.ArrayLike` annotation with `np.ndarray`.

**Comment by dineshbvadhia** [comment]
If the array type is known (eg. int, float) then use np.ndarray as the annotation instead of npt.ArrayLike - is that correct?

**Reply by BvB93** [reply]
> If the array type is known (eg. int, float)

Yes, assuming you're talking about the arrays' dtype here. 
In the current state of your example you're implying that the likes of `get_indices([0, 1, 2, 3])` _should_ be legal though via the use of `npt.ArrayLike`.
